hbase-server/src/main/java/org/apache/hadoop/hbase/security/access/SnapshotScannerHDFSAclHelper.java

```diff
@@ -0,0 +1,750 @@
+        .collect(Collectors.toList());
+  }
+
+  /**
+   * Return users with namespace read permission
+   * @param namespace the namespace
+   * @return users with namespace read permission
+   * @throws IOException if an error occurred
+   */
+  private Set<String> getUsersWithNamespaceReadAction(String namespace) throws IOException {
+    return PermissionStorage.getNamespacePermissions(conf, namespace).entries().stream()
+        .filter(entry -> entry.getValue().getPermission().implies(READ))
+        .map(entry -> entry.getKey()).collect(Collectors.toSet());
+  }
+
+  /**
+   * Return users with table read permission
+   * @param tableName the table
+   * @return users with table read permission
+   * @throws IOException if an error occurred
+   */
+  private Set<String> getUsersWithTableReadAction(TableName tableName) throws IOException {
+    return PermissionStorage.getTablePermissions(conf, tableName).entries().stream()
+        .filter(entry -> entry.getValue().getPermission().implies(READ))
+        .map(entry -> entry.getKey()).collect(Collectors.toSet());
+  }
+
+  private List<String> getNamespaceSnapshots(String namespace) throws IOException {
+    return getSnapshots((snapDesc) -> TableName.valueOf(snapDesc.getTable()).getNamespaceAsString()
+        .equals(namespace));
+  }
+
+  private List<String> getTableSnapshots(TableName tableName) throws IOException {
+    return getSnapshots((snapDesc) -> snapDesc.getTableName().equals(tableName));
+  }
+
+  private List<String> getSnapshots(Predicate<SnapshotDescription> predicate) throws IOException {
+    List<SnapshotDescription> snapshotDescriptions = admin.listSnapshots();
+    return snapshotDescriptions.stream()
+        .filter(snapshotDescription -> predicate.test(snapshotDescription))
+        .map(snapshotDescription -> snapshotDescription.getName()).collect(Collectors.toList());
+  }
+
+  protected PathHelper getPathHelper() {
+    return pathHelper;
+  }
+
+  /**
+   * Set HDFS acls
+   * @param hdfsAclOperations the {@link HDFSAclOperation} pair, the first HDFSAclOperation list
+   *          should be executed in a serial order, the second HDFSAclOperation list can be executed
+   *          concurrently.
+   */
+  private void setHDFSAcl(Pair<List<HDFSAclOperation>, List<HDFSAclOperation>> hdfsAclOperations)
+      throws InterruptedException, ExecutionException {
+    CompletableFuture<Void> future = setHDFSAclParallel(hdfsAclOperations.getSecond());
+    for (HDFSAclOperation hdfsAclOperation : hdfsAclOperations.getFirst()) {
+      setSingleHDFSAcl(hdfsAclOperation).get();
+    }
+    future.get();
+  }
+
+  private CompletableFuture<Void> setSingleHDFSAcl(HDFSAclOperation acl) {
+    return CompletableFuture.supplyAsync(() -> {
+      try {
+        acl.setAcl();
+      } catch (IOException e) {
+        LOG.error("Set HDFS acl error for path {}", acl.path.toString(), e);
+      }
+      return acl;
+    }, pool).thenComposeAsync(acl2 -> {
```

Comment:
So here,  we will submit the task twice ?  one for set acl and the other one for getChildrenOps ?    Actually, I'd prefer to make all those implementaion by using a limited produce-consume queue model....